### DISCUSSION TITLE
loop_function does not work

### DISCUSSION THREAD

**Discussion by melisa-akan** [jspsych/jsPsych#2999] [Q&A] [answered]
I cannot get the loop_function to work, even though I think I am using the exact solution offered[ here](https://github.com/jspsych/jsPsych/discussions/1408#discussioncomment-308793)

I am trying to loop through the same procedure for a given time period. I got rid of the irrelevant stuff to make the code simple. I hope I am providing enough info.  I tried to print a message on the console within the loop function but I do not see it. I guess it is not going into the loop function for some reason. 

Thank you for any guidance you can provide. 

```
var end_fam_time;

 var mt_instruction = {
	  type: jsPsychInstructions,
	  pages:  ['instructions here'],				
	  show_clickable_nav: true,
	  post_trial_gap: 1000,
	  on_start: function(){
		  end_fam_time = performance.now(); //this is where I start the timer
		   }
};

```

Below is the procedure I want to loop through. Everything works fine, except that it does not stop when the time is up.

```
mentalrot_procedure = {
  timeline: [mt_trials, feedback_mt],
  timeline_variables: mental_rot_stim,
   sample:{
    type:'custom',
    fn: function(t){
      let new_mt_order = [];
      let n_easy_trial = 8;
      let all_indeces = shuffle(t);
      
      for (let e = 0; e<n_easy_trial; e++){
        new_mt_order.push(e);
      }
      for(let n = 0; n<t.length; n++){
        new_mt_order.push(all_indeces[n]);
      }
      
      return new_mt_order;
    }
  }
  
};

//LOOP function

var loop_mt_trials = { //distractor task
  timeline: [mentalrot_procedure],
  loop_function: function(){
    console.log("HEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEY") // I do not see this in the console
  	let  curr_time = performance.now();
  	
  	if (curr_time - end_fam_time >=1*60*1000){ //should stop after 1 min but does not
  		return false;
  	}else{
  		return true;
  	}
  }
};

timeline.push(loop_mt_trials);



```

**Comment by jodeleeuw** [comment, marked as answer]
Hi @melisa-akan,

Is the `mentalrot_procedure` running as you expect? What happens when you get to the spot where it would loop? Does it just continue to the next part of the timeline?

**Reply by melisa-akan** [reply]
Thank you @jodeleeuw . Yes, it runs as I expect and I am now realizing that I did not title my inquiry correctly. It does loop; the  `mentalrot_procedure` runs repeatedly. It just does not get out of the loop when the time is up, it goes on forever. Sorry for the confusion. Thanks again!

**Reply by melisa-akan** [reply]
Sorry @jodeleeuw, I take it back. I think I was right about the loop function not working. (Nothing changes when I comment out  loop_function. It runs the trials defined in the  `mental_rot_stim` object. ) 

I don't have a good understanding of how the loop function is supposed to work when it is applied to a procedure that loops through the object defined in "timeline_variables"

**Reply by jodeleeuw** [reply]
The loop function will not run until all of the `timeline_variables` are completed, so I think the code is just not getting to the loop function yet. 

You'll have to move the `loop_function` onto the inner trial, and maybe use a different method than timeline variables to change the trials. You could, for example, create a global variable with an array of trials, and then use function parameters and a global index to track where you are in the set of trials. In the `loop_function` you can increment the index.

I think there are some examples of this kind of solution in the discussion forum somewhere, but I can't look for them at the moment.

**Reply by melisa-akan** [reply]
Thank you so so much @jodeleeuw ! This was super helpful. I solved my problem and putting the code below in case it helps anyone. 

My method here is not ideal as the elapsed time is checked only after the trials in the timeline are run. I wonder if there is a way to the check the passage of time independent of participant responses. If anyone can point me in the right direction to achieve this, I would be happy. But this solves my current issue. Thanks again for the super quick help. I really appreaciate it.


```
//create a custom index array inds

function createArray(n){
  let arr =[];
  for (let i=0; i<=n; i++){
    arr.push(i)
  }
  
  return arr;
}

var inds1 = createArray(8);
var inds2 = shuffle(createArray(383));
const inds = inds1.concat(inds2);
inds.push(inds);

var loop_index = 0;

var loop_mt_trials = { //distractor task
  timeline: [
    //test trial
    {type: jsPsychHtmlButtonResponse,
     stimulus:   function(){  
        let test_img = mental_rot_stim[inds[loop_index]].stim_name;
    		let html= "";
    		let style = `<style> .test-imgs {width: 240; height:200; margin:30px} </style>`;
    		html += style;
        html += `<div> <img src = "${test_img}.jpg" class = 'test-imgs'> </div>`;	
    		return html;
      },
    choices: ['Same','Different'],  
    post_trial_gap:0,
    data:{trial_type: "mentalrot_test",
    		    test_img:function(){

    		       return mental_rot_stim[inds[loop_index]].stim_name;
    		     },
    		     mt_trial: function(){
    		       return mental_rot_stim[inds[loop_index]].stim_name;
    		      }
          },
    on_finish: function(data){
      num_mt_trial += 1;
      data.accuracy = false;
      
      if(data.response==0 & mental_rot_stim[inds[loop_index]].CorAnswer=="Same"){
        data.accuracy = true;
        num_cor = num_cor + 1;
      }else if(data.response==1 & mental_rot_stim[inds[loop_index]].CorAnswer=="Dif"){
        data.accuracy= true;
        num_cor = num_cor + 1;
      }
      
      data.overall_mt_acc = num_cor/num_mt_trial;
      
      console.log("accuracy: "  + data.accuracy);
      console.log("trial: " + data.mt_trial);
      console.log("num_cor: " + num_cor);
      
     }
    },
    //feedback trial
    {type:jsPsychHtmlButtonResponse,
     stimulus:function(){
        var last_trial_correct = jsPsych.data.get().last(1).values()[0].accuracy;
        var overall_mt_acc = Math.round((jsPsych.data.get().last(1).values()[0].overall_mt_acc)*100);
        console.log("last_trial_correct: "  + last_trial_correct);
        
        if(last_trial_correct == 1){
          return "<p style='font-size:25px;color:#008000;'>Good job! <br><br> SCORE: " + overall_mt_acc + "% </p>"; // the parameter value has to be returned from the function
        } else {
          return "<p style='font-size:25px;color:#AA4A44;'>Incorrect. You'll do better next time. <br><br> SCORE: " + overall_mt_acc + "% </p>"; // the parameter value has to be returned from the function
        }
      },
      
      choices: ["NEXT"]
    }
    ],
    
  loop_function: function(){
       loop_index++;

  	let  curr_time = performance.now();
  	console.log("time_elapsed = " + (curr_time - end_fam_time));
  	
  	if (curr_time - end_fam_time >=1*60*1000){ //stop after 1 min 
  		return false;
  	}else{
  		return true;
  	}
  },
    
 on_finish: function(data){
   
   data.distractor_1_numT=num_mt_trial;
   data.distractor_1_acc=data.overall_mt_acc;


	}
};

timeline.push(loop_mt_trials);

```
